Optimize EmptyLineBetweenDefs

nelhage opened the pull request:
`Astrolabe::Node#sibling_index` scans the list of all siblings looking
for the target node, making it O(n) in the number of `def`s in a given
scope. This makes EmptyLineBetweenDefs quadratic in the size of a single
scope.

Instead, intercept `begin` nodes and walk the children explicitly,
keeping track of the previous node without having to do an O(n) scan to
find it.

alexdowad commented on lib/rubocop/cop/style/empty_line_between_defs.rb:
```diff
@@ -7,20 +7,26 @@ module Style
-        def on_method_def(node, _method_name, _args, _body)
-          return unless node.parent && node.parent.begin_type?
+        def on_begin(node)
+          prev = nil
```
You could make this a bit more concise using `Enumerable#each_cons`, rather than manually tracking `prev`.

alexdowad commented:
Aside from comment, looks good! Also need to fix failing build.

nelhage commented:
thanks! bumped to fix the build and use `each_cons`

bbatsov commented:
`Astrolabe`? We dropped that dependency recently.

bbatsov commented on lib/rubocop/cop/style/empty_line_between_defs.rb:
```diff
@@ -7,20 +7,21 @@ module Style
-          return unless node.parent && node.parent.begin_type?
-
-          nodes = [prev_node(node), node]
+        def on_begin(node)
```
I'd mention in the code why we've opted to do it like this, so someone wouldn't refactor it the way it was done before (which is pretty common in our codebase).

nelhage commented:
@bbatsov whoops, that was based on profiling an older version; Updated the commit message (the problem/optimization is the same with `RuboCop::Node#sibling_index`). Also added a comment.
